torch/distributed/_tensor/experimental/attention.py

```diff
@@ -361,153 +332,210 @@ def sdpa_backward_handler(
     assert output_sharding is not None, "output sharding should not be None"
     assert not output_sharding.needs_redistribute, "inputs need to be redistributed"
 
-    local_results = _scaled_dot_product_ring_flash_attention_backward(
-        op_info.mesh,
-        *op_info.local_args,  # type: ignore[arg-type]
-        **op_info.local_kwargs,  # type: ignore[arg-type]
-    )
+    if op_call == aten._scaled_dot_product_flash_attention_backward.default:
+        local_results = _scaled_dot_product_ring_flash_attention_backward(
+            op_info.mesh,
+            *op_info.local_args,  # type: ignore[arg-type]
+            **op_info.local_kwargs,  # type: ignore[arg-type]
+        )
+    elif op_call == aten._scaled_dot_product_efficient_attention_backward.default:
+        local_results = _scaled_dot_product_ring_efficient_attention_backward(
+            op_info.mesh,
+            *op_info.local_args,  # type: ignore[arg-type]
+            **op_info.local_kwargs,  # type: ignore[arg-type]
+        )
+    else:
+        raise NotImplementedError(f"{op_call=}")
 
     return DTensor._op_dispatcher.wrap(local_results, output_sharding.output_spec)
 
 
-def _scaled_dot_product_ring_flash_attention_backward(
+def _templated_ring_attention_backward(
     mesh: DeviceMesh,
+    op: AttentionOp,
     grad_out: torch.Tensor,
+    grad_out_name: str,
     query: torch.Tensor,
     key: torch.Tensor,
     value: torch.Tensor,
     out: torch.Tensor,
-    softmax_lse: torch.Tensor,
-    cum_seq_q: torch.Tensor,
-    cum_seq_k: torch.Tensor,
-    max_q: int,
-    max_k: int,
-    dropout_p: float,
+    logsumexp: torch.Tensor,
     is_causal: bool,
-    philox_seed: torch.Tensor,
-    philox_offset: torch.Tensor,
-    *,
-    scale: Optional[float] = None,
+    **kwargs: Any,
 ) -> Tuple[torch.Tensor, ...]:
     pg = mesh.get_group()
     assert isinstance(pg, dist.ProcessGroup), "must be single dimension"
     rank = dist.get_rank(pg)
     size = dist.get_world_size(pg)
-
-    # rank 0 sends to rank 1, rank 1 sends to rank 2, ..., rank n-1 sends to rank 0
-    right_dsts = list(range(1, size)) + [0]
-
     next_kv = None
+    next_grad_kv = None
+    rest: List[Any]
+    grad_query_, grad_key_, grad_value_ = None, None, None
 
-    out_grad_queries = []
-    out_grad_keys = []
-    out_grad_values = []
+    grad_query = torch.zeros_like(query, dtype=torch.float32)
```

Comment:
curious are the fp32 grad_query/key/value intentional?

Reply:
Yes, for the accuracy purpose. Otherwise, we may get NAN.